### PR TITLE
Fastroping

### PR DESCRIPTION
**Pull request by BaerMitUmlaut**
This is a port of AGMs fastroping and replaces #71. All functions have been completly rewritten, and pretty much only the base system remains the same. Includes a model from @KoffeinFlummi, slightly modified (added empty fire LOD).
Requires only two config entries from a mod maker to add the possibility to fast rope and can be extended with custom animations, FRIES etc.

Known issues:
- If a player is fastroping and the rope breaks below him (because it got shot at for example), he continues to fast rope just fine. This is because a broken rope cannot be retracted and it would look weird if he suddenly lets go of the rope, even though there is still a piece of rope below him.  
Maybe create a new, short rope, and retract that? Feel free to suggest a solution for this.
- You currently have to cut the ropes and redeploy the FRIES if both ropes break. That means you'll have to go through the *open door & extend hook* animation twice (which takes 8 seconds in the Ghosthawk). However, I think this is fair as you'd have to attach new ropes to the hooks, which would be hard to do with the hooks extended. IRL, a helo probably doesn't carry spare ropes either.
- The FRIES is not visible when you're sitting inside of the Ghosthawk. I think this is an LOD issue which I can do nothing about.

TODO:
- [ ] Do ? when bottom rope breaks
- [x] Add/Expand onPrepare function for all helos (e.g. KA-60 uses Czech animation names)
- [x] Add more FRIES models, texture current one - [reference for FRIES gantry](http://www.quantico.marines.mil/Portals/147/Docs/Range%20Management%20Branch/PubsInfo/3570/MCRP%203-11-4A.pdf), [reference for Chinook FRIES](https://www.flickr.com/photos/avgeekjoe/14390508271)
- [ ] Add function for automatic fast roping of AI (optional, but I've got the base of it already, might as well make some people happy :blush: )

### CODE REVIEW

**Comment by PabstMirror** [comment]
:fries: 
 I know a lot of people have been waiting for this.

**Comment by Glowbal** [comment]
Once you are done, could you clean up the commit history?

File: addons/fastroping/functions/fnc_canCutRopes.sqf [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * [_player, _vehicle] call ace_fastroping_canCutRopes
+ *
+ * Public: No
+ */
+
+#include "script_component.hpp"
+params ["_unit", "_vehicle"];
+private ["_deployedRopes"];
+
+_deployedRopes = _vehicle getVariable [QGVAR(deployedRopes), []];
```
Comment (Glowbal) [inline addons/fastroping/functions/fnc_canCutRopes.sqf]: `private _DeployedRopes =`

**Comment by BaerMitUmlaut** [comment]
Sure, will do.

File: addons/fastroping/functions/fnc_canCutRopes.sqf [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * Able to cut ropes <BOOL>
+ *
+ * Example:
+ * [_player, _vehicle] call ace_fastroping_canCutRopes
+ *
+ * Public: No
+ */
+
+#include "script_component.hpp"
+params ["_unit", "_vehicle"];
```
Comment (Glowbal) [inline addons/fastroping/functions/fnc_canCutRopes.sqf]: `_unit` doesn't seem to be used here

File: addons/fastroping/functions/fnc_canDeployRopes.sqf [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * Public: No
+ */
+
+#include "script_component.hpp"
+params ["_unit", "_vehicle"];
+private ["_deployedRopes", "_config"];
+
+_deployedRopes = _vehicle getVariable [QGVAR(deployedRopes), []];
+_config = configFile >> "CfgVehicles" >> typeOf _vehicle;
```
Comment (Glowbal) [inline addons/fastroping/functions/fnc_canDeployRopes.sqf]: `_config` is unused

File: addons/fastroping/functions/fnc_prepareFRIES.sqf [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+#include "script_component.hpp"
+params ["_unit", "_vehicle"];
+private ["_deployedRopes"];
+
+//Stage indicator: 0 - travel mode; 1 - preparing FRIES; 2 - FRIES ready; 3 - ropes deployed
+_vehicle setVariable [QGVAR(deploymentStage), 1, true];
+
+_config = configFile >> "CfgVehicles" >> typeOf _vehicle;
+_waitTime = 0;
```
Comment (Glowbal) [inline addons/fastroping/functions/fnc_prepareFRIES.sqf]: `private _waitTime =`

File: addons/fastroping/functions/fnc_prepareFRIES.sqf [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+#include "script_component.hpp"
+params ["_unit", "_vehicle"];
+private ["_deployedRopes"];
+
+//Stage indicator: 0 - travel mode; 1 - preparing FRIES; 2 - FRIES ready; 3 - ropes deployed
+_vehicle setVariable [QGVAR(deploymentStage), 1, true];
+
+_config = configFile >> "CfgVehicles" >> typeOf _vehicle;
```
Comment (Glowbal) [inline addons/fastroping/functions/fnc_prepareFRIES.sqf]: `private _config =`

**Comment by jonpas** [comment]
> The FRIES is not visible when you're sitting inside of the Ghosthawk. I think this is an LOD issue which I can do nothing about.

Yes, inside LOD rendering works differently, you can probably see it in some other position.

**Comment by BaerMitUmlaut** [comment]
It's fine in anything but first person view as a crew member.

**Comment by Khaosmatic** [comment]
Would it be possible to create SPIE Rigging from the same system?

File: addons/fastroping/functions/fnc_moduleEquipFRIES.sqf [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+ * Equips synched helicopters with a FRIES.
+ *
+ * Arguments:
+ * 0: Module <LOGIC>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [_vehicle] call ace_fastroping_fnc_cutRopes
```
Comment (bux578) [inline addons/fastroping/functions/fnc_moduleEquipFRIES.sqf]: wrong

**Comment by esteldunedain** [comment]
After the testing and fixing done yesterday, this should be ready for merging now. idk if it's feasable to do it for 3.5.0 or not, but I'm updating the milestone so it's reconsidered.

The bug @VKing6 happened when the invisible anchor at the end of the rope didn't touch the ground. That object is now removed.

**Comment by BaerMitUmlaut** [comment]
@Khaosmatic
>Would it be possible to create SPIE Rigging from the same system?

I'm not sure, especially because of the bug we found yesterday. If I had to guess I'd say no. Something like descending a stretcher or something similar from a helo would probably be possible, but attaching multiple objects or ropes to a rope seems to cause physics bugs.

**Comment by VKing6** [comment]
The ace_fastroping_helper is visible in the editor under empty->air as "Helicopter [Civilians]".

**Comment by BaerMitUmlaut** [comment]
Fixed.

**Comment by PabstMirror** [comment]
Mikro tools doesn't like something in config (`config_validator.py full`)
```
EQUIP_FRIES_ATTRIBUTE;
In File fastroping\CfgVehicles.hpp: Line 184 Expected Semicolon (or eol)
```
`x3 = "ace_fastropeing_helper" createVehicle [0, 0, 0];`
is `objNull` and rpt: Cannot create non-ai vehicle ace_fastropeing_helper,
(I don't think it can be scope =0 )??

**Comment by jonpas** [comment]
It has to be `scope = 1` to hide it in Editor but still allow usage with `createVehicle`.

**Comment by BaerMitUmlaut** [comment]
Fixed scopes and an error in the Eden macro. @PabstMirror can you run that check again? The script doesn't run for me (`AttributeError: 'module' object has no attribute 'run'`).

**Comment by PabstMirror** [comment]
Configs work now (I think the tool needs a configured p:, otherwise you can just run it though pboProject which will do same checks)

Vehicles create fine now. I did have the object name spelled wrong before :smile: , but I think the scope = 0 was still a problem.

**Comment by commy2** [comment]
merge this now ffs

**Pull request merged**